server/src/main/java/org/elasticsearch/timeseries/support/TimeSeriesMetrics.java

```diff
@@ -91,262 +120,409 @@
+            .map(d -> new TermsValuesSourceBuilder(d).field(d).missingBucket(true));
+        return new CompositeAggregationBuilder("time_series", sources.collect(toList())).aggregateAfter(afterKey).size(bucketBatchSize);
+    }
+
+    private CompositeAggregationBuilder timeSeriesComposite() {
+        Stream<CompositeValuesSourceBuilder<?>> sources = dimensionFieldNames.stream()
+            .map(d -> new TermsValuesSourceBuilder(d).field(d).missingBucket(true));
+        return new CompositeAggregationBuilder("time_series", sources.collect(toList())).size(bucketBatchSize);
+    }
+
+    private AggregationBuilder latestMetric(String metric, long time, TimeValue range, TimeValue step) {
+        if (step == null) {
+            String aggKey = range == null ? Long.toString(time) : "use_timestamp";
+            return new FilterAggregationBuilder(metric, new ExistsQueryBuilder(metric)).subAggregation(
+                new FilterAggregationBuilder(aggKey, new MatchAllQueryBuilder()).subAggregation(
+                    new TopHitsAggregationBuilder("results").sort(
+                        new FieldSortBuilder("@timestamp").order(range == null ? SortOrder.DESC : SortOrder.ASC)
+                    )
+                        .fetchField(metric)
+                        .fetchField(new FieldAndFormat("@timestamp", "epoch_millis"))
+                        .size(range == null ? 1 : MAX_INNER_RESULT_WINDOW_SETTING.getDefault(Settings.EMPTY))
+                )
+            );
+        } else {
+            FilterAggregationBuilder metricAgg = new FilterAggregationBuilder(metric, new ExistsQueryBuilder(metric));
+            long stepMillis = step.getMillis();
+            long from = time - range.getMillis();
+            long first = from / stepMillis;
+            long last = time / stepMillis;
+            FilterAggregationBuilder[] filters = new FilterAggregationBuilder[(int) (last - first)];
+            for (int i = 0; i < filters.length; i++) {
+                metricAgg.subAggregation(
+                    new FilterAggregationBuilder(
+                        Long.toString((first + i + 1) * stepMillis),
+                        new BoolQueryBuilder().filter(
+                            new RangeQueryBuilder("@timestamp").format(DateFieldMapper.DEFAULT_DATE_TIME_NANOS_FORMATTER.pattern())
+                                .gt((first + i + 1) * stepMillis - staleness.getMillis())
+                                .lte((first + i + 1) * stepMillis)
+                        ).filter(new ExistsQueryBuilder(metric))
+                    ).subAggregation(
+                        new TopHitsAggregationBuilder("results").sort(new FieldSortBuilder("@timestamp").order(SortOrder.DESC))
+                            .fetchField(metric)
+                            .fetchField(new FieldAndFormat("@timestamp", "epoch_millis"))
+                            .size(1)
+                    )
+                );
             }
+            return metricAgg;
```

Comment:
This whole thing is going to be really fun to run. I get it, I think, but wow. I guess this is what we need.

Reply:
We might be able to improve things by having top_metrics have a better behavior when the field is missing. But that might be a thing for another time.